zkevm-circuits/src/evm_circuit/execution/stop.rs

```diff
@@ -21,28 +34,192 @@ impl<F: Field> ExecutionGadget<F> for StopGadget<F> {
     const EXECUTION_STATE: ExecutionState = ExecutionState::STOP;
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let code_size = cb.bytecode_length(cb.curr.state.code_hash.expr());
+        let is_out_of_range =
+            IsZeroGadget::construct(cb, code_size.expr() - cb.curr.state.program_counter.expr());
         let opcode = cb.query_cell();
-        cb.opcode_lookup(opcode.expr(), 1.expr());
+        cb.condition(1.expr() - is_out_of_range.expr(), |cb| {
+            cb.opcode_lookup(opcode.expr(), 1.expr());
+        });
 
-        // Other constraints are ignored now for STOP to serve as a mocking
-        // terminator
+        // We do the responsible opcode check explicitly here because we're not using
+        // the `SameContextGadget` for `STOP`.
+        cb.require_equal(
+            "Opcode should be STOP",
+            opcode.expr(),
+            OpcodeId::STOP.expr(),
+        );
 
-        Self { opcode }
+        // Call ends with STOP must be successful
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 1.expr());
+
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr() + is_to_end_tx.clone(),
+            2.expr() * cb.curr.state.is_root.expr() * is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // When a transaction ends with STOP, this call must be persistent
+            cb.call_context_lookup(
+                false.expr(),
+                None,
+                CallContextFieldTag::IsPersistent,
+                1.expr(),
+            );
+
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(2.expr()),
+                ..StepStateTransition::any()
+            });
+        });
+
+        // When it's an internal call
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(cb, 1.expr(), 0.expr(), 0.expr())
+        });
+
+        Self {
+            code_size,
+            is_out_of_range,
+            opcode,
+            restore_context,
+        }
     }
 
     fn assign_exec_step(
         &self,
         region: &mut CachedRegion<'_, '_, F>,
         offset: usize,
-        _: &Block<F>,
+        block: &Block<F>,
         _: &Transaction,
-        _: &Call,
+        call: &Call,
         step: &ExecStep,
     ) -> Result<(), Error> {
+        let code = block
+            .bytecodes
+            .iter()
+            .find(|b| b.hash == call.code_hash)
```

Comment:
Hm, I think we should use a hashmap here to avoid this iteration.  But we can leave this for a future PR!

Reply:
It seems not much diff to do, so I updated it in [`b6e0a73`](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/419/commits/b6e0a73f559b356117f9d144c6d0f2159e460d2a) and it looks better now!